### PR TITLE
Added the sparse simulator sample.

### PR DESCRIPTION
**Pull request by kuzminrobin**

### CODE REVIEW

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: This sample looks to be in the wrong folder? `/samples/azure-quantum` is intended for service-oriented samples, but this is all about local usage.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:10]: A few comments on this cell:

- Need something higher-level as an intro, since portal gallery samples may be discoverable by users new to quantum. The sample on the whole need not be beginner friendly, but the introduction should at least clearly set the context and expectations, and redirect to other resources if a user wants to exit to other learning paths.
- > The sparse simulator is efficient for a class of quantum algorithms with a smaller number of states in superposition.
  
  This isn't terribly precise as written — it would be more correct, I think, to note that the sparse simulator is efficient for representing quantum states that are sparse in the computational basis.
- > This simulator is an unlocker for users to explore larger applications than what can be explored using the full state simulator alone.

  This isn't quite gramatically correct, and is a bit tech-jargony. I would suggest something along these lines instead:

  > This simulator is useful for exploring larger applications than can be represented using dense simulation alone.

- > For more information about the sparse simulator see [this paper](https://arxiv.org/abs/2105.01533).

  For both accessibility and SEO, links should be descriptive. Please avoid "this paper" as link text. Here, I would suggest:
  
  > For more information about the sparse simulator, please see [Jaques and Häner (arXiv:2105.01533)](https://arxiv.org/abs/2105.01533).

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:12]:
```suggestion
    "In this sample, we'll look at the advantages of the sparse simulator over the [full-state simulator](https://docs.microsoft.com/azure/quantum/user-guide/machines/full-state-simulator).\n",
```

Please note that `en-us` in docs URLs overrides users' locale preferences and should always be removed from links.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:14]:
```suggestion
    "Below, we'll create a few clusters of qubits, each prepared in a GHZ state. To run the cell below, press Ctrl+Enter."
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:40]: These two are implied in IQ# kernels and should be removed.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:43]: As is, the notebook jumps right in without explaining any of the code or how it's used. Please break these into a separate cell, and explain what these namespaces provide for user code.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: Please make sure to make this notebook available in other samples experiences — the portal experience is only one of our four samples experiences, such that this notebook must be added to `main` as well before the sample can be considered complete. Thanks!

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:45]: These comments should be moved into documentation comments, as that also teaches users how to document their own Q# code.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:57]: This could probably be simplified by iterating over the output of `Microsoft.Quantum.Arrays.Chunks`.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:61]:
```suggestion
    "        Message($\"Finished preparing state for cluster {i / clusterSize}.\");\n",
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:70]: `MultiM` has been deprecated in favor of `M` and `ForEach`:

```suggestion
    "    let results = ForEach(M, qs);\n",
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:85]:
```suggestion
    "Let's make the small states displayed conveniently on a histogram. Please execute the cell below."
```

Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:85]: Similar here as well about "Please execute..." It would be good to follow the prose style for other notebook samples.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:107]: Linking to the docs on visualization configuration may be helpful here.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2180]:
```suggestion
    "In the cell above increase the `nQubits` input (by multiples of `nClusters` parameter) such that the simulation becomes slow (e.g. `nQubits`: 8, 10, 12, 14, ...; `nClusters`: 2). On some laptops the simulation becomes slow at `nQubits=20`. Don't make large steps otherwise the simulation can run too long.\n",
```

Other notes:
- If this is intended for use in the portal gallery as well, "laptop" doesn't have a lot of meaning here.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2181]:
```suggestion
    "After the number of qubits exceeds 15 the state histogram will not be displayed."
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2189]:
```suggestion
    "Now let's run our Q# operation on the sparse simulator, using the [%simulate_sparse](https://docs.microsoft.com/qsharp/api/iqsharp-magic/simulate_sparse) magic command."
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2236]:
```suggestion
    "From the histogram above you can see that the sparse simulator only stores non-zero amplitudes, as opposed to the full-state simulator.\n",
```

Technically, this isn't _quite_ true, since the visualization functionality in IQ# allows truncating out amplitudes that are close to zero when printing from the dense simulator as well. It may be good to clarify slightly.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2238]:
```suggestion
    "By increasing the `nQubits` parameter you can see that the sparse simulator still provides the result quickly enough, even for registers large enough that the full state simulator slows down.\n",
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2239]:
```suggestion
    "You can increase the number of qubits up to 1,024, which is the current hardcoded limit in the sparse simulator."
```

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [inline samples/azure-quantum/simulators/SparseSimulator.ipynb:2247]: I'd again suggest rephrasing this cell in a user-focused way — "unlocker" is kind of tech jargon and isn't common for users that will be reading this sample.

**Review by cgranade — CHANGES_REQUESTED** [review]
I think this is looking like a strong start, but will need some more iteration before we can call the sample complete. Thanks for opening this PR!

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: What would be the right folder for the `SparseSimulator.ipynb`?

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:10]: Updating.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: DevOps item 39303.
I assume the folder in `main` will be the same as [in AQ gallery branch](https://github.com/microsoft/Quantum/pull/637#discussion_r839937259). Otherwise please specify the folder in `main`.

Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: I'm not quite sure, let me think. In general, we haven't had samples decidated to individual simulators, so much as how to use those simulators in more user-focused samples. I'd honestly prefer that approach here as well, but I understand that design constraints preclude that. Perhaps the right approach is to rename the sample to something like "simulating quantum programs with large registers," and place in `getting-started/`?

Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: https://github.com/microsoft/Quantum/issues/639
I assume that the folder in `main` will be the same as [in the AQ gallery branch](https://github.com/microsoft/Quantum/pull/637#discussion_r839937259), otherwise please specify the folder in `main`.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:40]: What do you think about commenting these lines out, but adding a comment on the right `// Required in stand-alone.`?
```q#
// open Microsoft.Quantum.Canon;       // Required in stand-alone.
// open Microsoft.Quantum.Intrinsic;   // Required in stand-alone.
```

Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:40]: I would strongly advise against doing so — this sample documents using the feature from a Q# standalone notebook, not from Q# standalone CLI. A user should not need to know how to use Q# to write standalone CLI applications in order to take advantage of this notebook.

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:12]: We should be consistent, to be sure. "Full state" strikes me as incorrect grammatically, but at the same time I don't think we should go and fix everywhere else that occurs at the moment. @SoniaLopezBravo, do you have any advice here from a docs perspective?

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:1]: Please see comment above. Thanks!

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb:2239]: Changing to 
> You can increase the number of qubits up to **1,024. That** is the current hardcoded limit in the sparse simulator.

(comma after `1,024` encourages a misleading impression for me, as if 1 and 024 were separate numbers in an enumeration, or were limits of a range)

**Comment by kuzminrobin** [comment]
The first-most link (to the sparse simulator doc) is not live yet. Will go live on Monday at 10am.
Other than that all the feedback is addressed and the PR looks ready to be merged.

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:49]: This may be hard to read on narrower screens; I'd suggest moving the comments out to a text cell before this code cell. Perhaps something along these lines?

> Below, we'll open a few Q# namespaces to provide access to functions for creating manipulating arrays of data, and for helping to visualize our quantum programs. To learn more, check out the [sample on visualizing quantum programs](https://github.com/microsoft/Quantum/tree/main/samples/diagnostics/visualization) and the [Q# API reference](https://docs.microsoft.com/qsharp/api/qsharp/).

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:80]:
```suggestion
    "/// Number of qubits to allocate (must be a multiple of the nClusters input).\n",
```

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:91]: How readable are these comments on narrower screens? It's also worth keeping in mind that in the Azure portal, other blades can claim a fair bit of horizontal space. It may be worth moving these comments to be above relevant lines instead of to the side.

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:121]:
```suggestion
    "For more information about `dump.basisStateLabelingConvention`, please see [Configuration settings](https://docs.microsoft.com/qsharp/api/iqsharp-magic/config#configuration-settings)."
```

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:2214]:
```suggestion
    "In the cell above increase the `nQubits` input (by multiples of `nClusters`) such that the simulation becomes slow (e.g. `nQubits`: 8, 10, 12, 14, ...; `nClusters`: 2). The simulation is expected to become slow at `nQubits=24` and to take about a minute with `nQubits=26`. It is not recommended to make large steps, otherwise the simulation can run too long.  \n",
```

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:2215]:
```suggestion
    "After the number of qubits exceeds 15, the state histogram will not be displayed."
```

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (cgranade) [inline samples/getting-started/simulation/LargeSimulation.ipynb:2440]: Are there any CTAs we want to provide for users to continue through to?

**Review by cgranade — COMMENTED** [review]
This is looking a lot closer, thanks for all your revisions! I had just a few last comments to go, then we should be able to merge this in and start the cherry-pick onto `main` as well. Thank you!

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (kuzminrobin) [reply samples/getting-started/simulation/LargeSimulation.ipynb:121]: Fixing. I apologize for having you to report this multiple times.

Comment (cgranade) [reply samples/getting-started/simulation/LargeSimulation.ipynb:121]: No worries at all, it's easy to miss!

File: samples/getting-started/simulation/LargeSimulation.ipynb
Comment (kuzminrobin) [reply samples/getting-started/simulation/LargeSimulation.ipynb:2440]: Adding:
## See Also
* [Quantum simulators](https://docs.microsoft.com/en-us/azure/quantum/user-guide/machines/)
* [Q# API reference](https://docs.microsoft.com/qsharp/api/qsharp/)

I also plan to add a link to [Mathias' sample](https://github.com/microsoft/Quantum/pull/619) of integer factorization with SS, once that sample is merged.
Any other good links you have in mind?

File: samples/azure-quantum/simulators/SparseSimulator.ipynb
Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb]: For now I'm leaving dashes (`full-state`).
And on my side I'll try to clarify with the rest of the team.

Comment (cgranade) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb]: It still seems really wrong to me, but if we're using "full state" consistently elsewhere, we should do so here as well.

Comment (kuzminrobin) [reply samples/azure-quantum/simulators/SparseSimulator.ipynb]: Changed to "full state".

**Comment by kuzminrobin** [comment]
I believe all the feedback has been addressed.
Please unblock/approve or provide more feedback.

**Review by cgranade — APPROVED** [review]

**Pull request merged**
